src/permissions.ts

```diff
@@ -89,3 +94,88 @@ export async function checkPermissions<T>(
   }
   return anyTrue;
 }
+
+// Check the permissions for a given model and method. If the method is a read, update, or delete,
+// finds the relevant object, checks the permissions, and attaches the object to the request as
+// req.obj.
+export function permissionMiddleware<T>(
+  baseModel: Model<T>,
+  options: Pick<FernsRouterOptions<T>, "permissions" | "populatePaths" | "discriminatorKey">
+) {
+  return async (req: express.Request, res: express.Response, next: NextFunction) => {
+    if (req.method === "OPTIONS") {
+      return next();
+    }
+    try {
+      let method: "list" | "create" | "read" | "update" | "delete";
+
+      const reqMethod = req.method.toLowerCase();
+      if (reqMethod === "post") {
+        method = "create";
+      } else if (reqMethod === "get") {
+        if (req.params.id) {
+          method = "read";
+        } else {
+          method = "list";
+        }
+      } else if (reqMethod === "patch") {
+        method = "update";
+      } else if (reqMethod === "delete") {
+        method = "delete";
+      } else {
+        throw new APIError({
+          status: 405,
+          title: `Method ${req.method} not allowed`,
+        });
+      }
+
+      const model = getModel(baseModel, req.body, options);
+
+      // All methods check for permissions.
+      if (!(await checkPermissions(method, options.permissions[method], req.user))) {
+        throw new APIError({
+          status: 405,
```

Comment:
Since this is a permissions thing, I'd go with 403 here.

Reply:
Ah I think 405 is appropriate here and gives more context (you can't do _anything_ with this endpoint, not you can't do something with this specific item). I got the inspiration for this from Django REST Framework and it has helped me in the past (instead of two different ways you could get a 403 here).

Reply:
Nope you were totally right here.